### PR TITLE
refactor(meta): Ensure all moving values remain unchanged between two transactions

### PR DESCRIPTION
**Pull request by WenyXu**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## What's changed and what's your 

1. Introduce `move_values` in `TombstoneManager`
2. Ensure all moving values remain unchanged between two transactions
3. Make `delete_table_metadata` and `restore_table_metadata` to be Idempotent.

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.
- [x]  This PR does not require documentation updates.

### CODE REVIEW

**Review by tisonkun — COMMENTED** [review]
What is the purpose of this refactor?

**Comment by WenyXu** [comment]
> What is the purpose of this refactor?

For moving value to tombstone values operation, it ensures all values remain unchanged between two transactions instead of a single value in the previous version.

File: src/common/meta/src/key.rs [hunk 1]
```diff
@@ -666,7 +659,7 @@ impl TableMetadataManager {
         let operations = self
             .table_metadata_keys(table_id, table_name, table_route_value)?
             .into_iter()
-            .map(|key| TxnOp::Delete(key.into_bytes()))
+            .map(TxnOp::Delete)
```
Comment (fengjiachun) [inline src/common/meta/src/key.rs:662]: Better to use [batch_delete](https://github.com/GreptimeTeam/greptimedb/blob/09b3c7029b7d7ac8abd1cebb613c9a00ea7c66bb/src/common/meta/src/kv_backend.rs#L65), it has no keys count limit.

File: src/common/meta/src/key/tombstone.rs [hunk 2]
```diff
@@ -29,307 +29,158 @@ pub(crate) struct TombstoneManager {
...
-        impl FnMut(&'_ mut TxnOpGetResponseSet) -> Option<Vec<u8>>,
-    ) {
-        TxnOpGetResponseSet::build_get_op(to_tombstone(self.0))
+        src_key: Vec<u8>,
+        value: Vec<u8>,
+        dest_key: Vec<u8>,
+    ) -> (Txn, impl FnMut(&mut TxnOpGetResponseSet) -> Option<Vec<u8>>) {
+        let txn = Txn::new()
+            .when(vec![
+                Compare::with_not_exist_value(dest_key.clone(), CompareOp::Equal),
```
Comment (fengjiachun) [inline src/common/meta/src/key/tombstone.rs:63]: Do not need compare `dest_key`

File: src/common/meta/src/key/tombstone.rs [hunk 3]
```diff
@@ -29,307 +29,158 @@ pub(crate) struct TombstoneManager {
...
-    /// Returns a new normal key.
-    pub(crate) fn new<T: Into<Vec<u8>>>(key: T) -> Self {
-        Self {
-            bytes: key.into(),
-            atomic: false,
+        error::MoveValuesSnafu {
+            err_msg: format!(
+                "keys: {:?}, dest keys: {:?}",
+                keys.iter().map(|key| String::from_utf8_lossy(key)),
+                dest_keys.iter().map(|key| String::from_utf8_lossy(key))
```
Comment (fengjiachun) [inline src/common/meta/src/key/tombstone.rs:121]: Do not need to give `dest_keys` since we can obtain them from the `keys`.

**Review by fengjiachun — APPROVED** [review]
LGTM

File: src/common/meta/src/rpc/store.rs [hunk 4]
```diff
@@ -509,6 +509,13 @@ impl BatchDeleteRequest {
         }
     }
 
+    #[inline]
+    /// Sets `keys`.
```
Comment (tisonkun) [inline src/common/meta/src/rpc/store.rs:513]: This should violate the latest style guide to hold comment above attributions.

And IIRC @killme2008 suggests that we never use `#[inline]` but let the compiler decide it.

File: src/common/meta/src/rpc/store.rs [hunk 5]
```diff
@@ -509,7 +509,6 @@ impl BatchDeleteRequest {
         }
     }
 
-    #[inline]
```
Comment (fengjiachun) [inline src/common/meta/src/rpc/store.rs:512]: remove all `inline` in this file.

File: src/common/meta/src/key/tombstone.rs [hunk 6]
```diff
@@ -29,307 +29,154 @@ pub(crate) struct TombstoneManager {
...
+                    );
+                    (txn, (key.clone(), filter))
+                })
+                .unzip();
+            let mut resp = self.kv_backend.txn(Txn::merge_all(txns)).await?;
+            if resp.succeeded {
+                return Ok(());
+            }
+            let mut set = TxnOpGetResponseSet::from(&mut resp.responses);
+            // Updates results.
```
[the remark this answers is not in the document]
Comment (WenyXu) [reply src/common/meta/src/key/tombstone.rs:106]: No, it only retrieves the latest values and starts a new transaction.

**Review by MichaelScofield — APPROVED** [review]

**Pull request merged**